src/librustc/hir/mod.rs

```diff
@@ -961,22 +956,24 @@ pub enum Expr_ {
     ExprRepeat(P<Expr>, P<Expr>),
 }
 
-/// The explicit Self type in a "qualified path". The actual
-/// path, including the trait and the associated item, is stored
-/// separately. `position` represents the index of the associated
-/// item qualified with this Self type.
-///
-///     <HirVec<T> as a::b::Trait>::AssociatedItem
-///      ^~~~~     ~~~~~~~~~~~~~~^
-///      ty        position = 3
-///
-///     <HirVec<T>>::AssociatedItem
-///      ^~~~~    ^
-///      ty       position = 0
+/// Optionally `Self`-qualified value/type path or associated extension.
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
-pub struct QSelf {
-    pub ty: P<Ty>,
-    pub position: usize,
+pub enum QPath {
```

Comment:
I'll have to see how this structure works out -- it doesn't seem like what I would naively expect, which is probably something more like:

```rust
pub enum Path {
    Base(P<PathBase>),
    Project(P<Path>, P<PathProjection>),
}

pub enum PathBase {
    Module(P<Path>), // i.e., `::a` (or `a`, in a `use` statement)
    FullyQualified(P<Ty>, P<Path>), // i.e., `<Ty as Path>`
    Type(P<Ty>),
}

...
```

But I can see that the first two cases are kind of the same as `QPath` (also, the structure that I gave is less precise, in that it would permit associated modules etc). But anyway the point is some kind of recursive tree.

That said, I am also happy with this `QPath` structure -- I think mostly I just want some better documentation. Specifically, some examples showing different paths and how they wind up being represented to make it clearer.